server/src/main/java/org/elasticsearch/features/FeatureService.java

```diff
@@ -0,0 +1,236 @@
+        readSpecs();
+
+        // use sorted sets here so they're easier to look through in info & debug output
+        NavigableMap<Version, NavigableSet<String>> declaredHistoricalFeatures = new TreeMap<>();
+        for (FeatureSpecification spec : FEATURE_SPECS.values()) {
+            for (var f : spec.getHistoricalFeatures().entrySet()) {
+                if (FeatureEra.isPublishable(f.getValue().major) == false) continue;    // don't include, it's before the valid eras
+                declaredHistoricalFeatures.computeIfAbsent(f.getValue(), k -> new TreeSet<>()).add(f.getKey().id());
+            }
+        }
+
+        // add in all features from previous versions
+        Set<String> featureAggregator = new TreeSet<>();
+        for (Map.Entry<Version, NavigableSet<String>> versions : declaredHistoricalFeatures.entrySet()) {
+            featureAggregator.addAll(versions.getValue());
+            versions.setValue(Collections.unmodifiableNavigableSet(new TreeSet<>(featureAggregator)));
+        }
+
+        return Collections.unmodifiableNavigableMap(declaredHistoricalFeatures);
+    }
+
+    /**
+     * Returns the features that are implied by a node of {@code version}.
+     */
+    static Set<String> readHistoricalFeatures(Version version) {
+        var aggregates = HISTORICAL_FEATURES;
+        if (aggregates == null) {
+            aggregates = HISTORICAL_FEATURES = calculateHistoricalFeaturesFromSpecs();
+        }
+        var features = aggregates.floorEntry(version);
+        return features != null ? features.getValue() : Set.of();
+    }
+
+    private static Set<String> calculateFeaturesFromSpecs() {
+        readSpecs();
+
+        Set<String> features = new TreeSet<>();
+        for (FeatureSpecification spec : FEATURE_SPECS.values()) {
+            for (NodeFeature f : spec.getFeatures()) {
+                if (f.era().isPublishable()) {
+                    features.add(f.id());
+                }
+            }
+        }
+        return Collections.unmodifiableSet(features);
+    }
+
+    /**
+     * Returns all the features registered by all loaded {@link FeatureSpecification} instances
+     */
+    public static Set<String> readFeatures() {
+        var features = NODE_FEATURES;
+        if (features == null) {
+            features = NODE_FEATURES = calculateFeaturesFromSpecs();
+        }
+        return features;
+    }
+
+    private final Set<String> features;
+
+    /**
+     * Construct a new {@code FeatureService} for the local node, using all available feature specifications.
+     */
+    public FeatureService() {
+        features = readFeatures();
+    }
+
+    /**
+     * Construct a new {@code FeatureService} to expose {@code features}
+     */
+    public FeatureService(Set<String> features) {
+        this.features = Set.copyOf(features);
+    }
+
+    /**
+     * Returns the set of features published by this node
+     */
+    public Set<String> readPublishableFeatures() {
```

Comment:
Nit: since this became just a getter, maybe rename it to `getFeatures()` or `getPublishableFeatures()`?